src/main/java/com/terraformersmc/terrestria/config/TerrestriaConfigManager.java

```diff
@@ -3,140 +3,132 @@
+			byte[] bytes = Files.readAllBytes(configPath);
 
-				generalConfig = GSON.fromJson(br, TerrestriaGeneralConfig.class);
-			}
-		} catch (FileNotFoundException e) {
-			System.err.println("Couldn't load Terrestria general configuration file; reverting to defaults");
-			e.printStackTrace();
-		}
-	}
+			// Make sure that we're using UTF-8. This prevents weird locale / OS issues.
+			content = new String(bytes, StandardCharsets.UTF_8);
+		} catch (IOException e) {
+			Terrestria.LOGGER.error("Failed to load Terrestria configuration file at " + configPath, e);
+			Terrestria.LOGGER.error("This shouldn't happen under normal conditions, ensure that you have the correct permissions");
+			Terrestria.LOGGER.error("Reverting to default configuration");
 
-	public static void saveClient() {
-		if (FabricLoader.getInstance().getEnvironmentType() != EnvType.CLIENT) {
-			throw new UnsupportedOperationException("Cannot save the Terrestria client configuration when not on the client.");
+			return defaults;
 		}
-		prepareFiles();
 
-		String jsonString = GSON.toJson(clientConfig);
+		try {
+			return GSON.fromJson(content, clazz);
+		} catch (JsonSyntaxException e) {
+			Terrestria.LOGGER.error("Failed to parse Terrestria configuration file at " + configPath, e);
+
+			// Revert the config so that the user has a fresh start if they need it
+			// It's also possible for the user to delete the config to recreate it, but that seems a bit unintuitive
+			Terrestria.LOGGER.error("Reverting to default configuration, ensure that your file has correct syntax");
+			saveConfig(configPath, defaults);
+
+			// There are a few websites like this, but this was the first result that came up.
+			Terrestria.LOGGER.error("In the future, consider using something like https://jsonchecker.com/ to check your syntax");
+
+			// It would be quite annoying if a user just spent 10 minutes editing the file, only for it to be wiped away.
+			// Save a backup so they can review the errors and fix them.
+			Terrestria.LOGGER.error("The previous configuration file content has been written to " + backupPath);
+
+			try {
+				Files.write(backupPath, content.getBytes(StandardCharsets.UTF_8));
+			} catch (IOException ioe) {
+				Terrestria.LOGGER.error("Couldn't save previous configuration file content at " + backupPath, ioe);
+				Terrestria.LOGGER.error("This shouldn't happen under normal conditions, ensure that you have the correct permissions");
+			}
 
-		try (FileWriter fileWriter = new FileWriter(clientFile)) {
-			fileWriter.write(jsonString);
-		} catch (IOException e) {
-			System.err.println("Couldn't save Terrestria client configuration file");
-			e.printStackTrace();
+			return defaults;
 		}
 	}
 
-	public static void saveGeneral() {
-		prepareFiles();
-
-		String jsonString = GSON.toJson(generalConfig);
+	private static <T> void saveConfig(Path configPath, T instance) {
+		String jsonString = GSON.toJson(instance);
```

Comment:
Just making sure that prettyPrinting doesn't have to be enabled here to make the configuration use multiple lines

Reply:
I just checked: If prettyPrinting is not used, then the whole file is stored in a single line. I think that this probably isn't what we want.

Reply:
Super easy to do:
```
Gson gson = new GsonBuilder().setPrettyPrinting().create();
 String json = gson.toJson(lang);
```

Reply:
Well, that GSON instance already has pretty printing enabled. I thought you just wanted to check whether it was necessary.

```java
private static final Gson GSON = new GsonBuilder().setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES).setPrettyPrinting().create();
```

Reply:
Ahh I missed that part, I was making sure it was enabled, so we're good on that front then.